src/Workspaces/Core/Portable/Formatting/Formatter.cs

```diff
@@ -146,6 +146,14 @@ internal static async Task<Document> FormatAsync(Document document, SyntaxAnnota
             }
 
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Languages such as Razor can specify their own options instead of using the document's options.
+            var documentOptionSetProvider = document.Services.GetService<IDocumentOptionSetProvider>();
+            if (documentOptionSetProvider is not null)
+            {
+                options = await documentOptionSetProvider.GetOptionsForDocumentAsync(document, cancellationToken).ConfigureAwait(false);
```

Comment:
Hmmm, one interesting implication is that if you pass in options to this method they'll be ignored because the document options will win. With that in mind the options on line 157 account for this; in an ideal world all call sites to this method wouldn't actually pass options in "within" Roslyn and would then fall back to the appropriate document options services. All that can be done in the future though, just my 2cents

Reply:
Yeah this seems really weird here. Couldn't the LSP layer or something deal with coupling this at that layer versus something else?

Reply:
Currently Razor is the only implementer of `IDocumentSetProvider` so in practice I don't think there's currently an issue, however I totally see both of your points about the check being weird here. I'm exploring alternate options, will hold off on merging in the meantime.